src/hotspot/share/logging/logConfiguration.cpp

```diff
@@ -253,19 +253,44 @@ void LogConfiguration::configure_output(size_t idx, const LogSelectionList& sele
     on_level[level]++;
   }
 
+  // MT-SAFETY
+  //
+  // ConfigurationLock can only guarantee that only one thread is doing reconfiguration. This function still needs
+  // to be MT-safe because logsites in other threads may be executing in parallel. Reconfiguration means unified
+  // logging allows users to dynamically change tags and decorators of a log output via DCMD.
+  //
+  // A RCU-style synchronization 'wait_until_no_readers()' is imposed inside of 'ts->set_output_level(output, level)'
+  // above if setting has changed. It guarantees that all logs either synchronous writing or enqueuing to the async
+  // buffer see the new tags and decorators. It's worth noting that the synchronization happens even level does not
+  // change.
+  //
+  // LogDecorator is a set of decorators represented in a uint. sizeof(uint) is not greater than a machine word,
+  // so store of it is atomic on the mainstream processors. I.e. readers see either its older value or new value.
```

Comment:
This is not necessarily true - the field also has to have correct alignment to ensure atomic accesses. The convention we are adopting is that all racily accessed fields like _decorators should be declared volatile, and accessed using Atomic::load and Atomic::store.

Reply:
To support machines without unaligned memory access, C++ defines "alignment requirement" for class and struct. 
Not only sizeof(LogDecorators) == sizeof(uint) == 4, C++ can also guarantee that alignof(logDecorators) is 4. 
Therefore, it's impossible that to unaligned access for its field _decorators. 

```
class LogDecorators {
  uint _decorators;
};
```

Therefore, I think the current code which use normal assignment is atomic. Of course, using Atomic::store manifests it's an atomic store. I will try to use Atomic::store().

Reply:
Could you take a look at my argument? IMHO, C++ can guarantee the following code is atomic. 
Alignment shouldn't an issue. 

```
void LogOutput::set_decorators(const LogDecorators& decorators) {
    _decorators = decorators;
}
```

I manage to support the following code. The generated code is exactly same, but it needs some tricks of c++ meta-programming.  Is it worthy?  if you still insist we should use Atomic::store, how about I create a separate JBS issue and submit a patch?  My concern is that Atomic::load/store makes current code messy. 

```
void LogOutput::set_decorators(const LogDecorators& decorators) {
  Atomic::store(&_decorators, decorators);
}
```

Reply:
C++ makes no guarantees about atomicity unless using C++ atomic operations. Yes they probably are atomic but we don't rely on "probably". If a variable is being accessed lock-free and can take part in data-race then please just use Atomic::load and store for simple accesses (assuming they are correct in the context of how the variable can be used concurrently).

Reply:
Sorry I just realized what you are saying here. This is a LogDecorator type not a primitive type so the notion of atomic load/store doesn't make sense. The comments about it being a bitfield threw me.

Reply:
It is the load/store of `uint _decorators;` that needs to be atomic.

Reply:
First my apologies for getting muddled by this code. We don't use class types in lock-free algorithms so don't have a means to express atomic updates to objects this way. We still need the accesses to be atomic, but have no way to express that in the source code when the code has been written this way.

The issue is not whether the generated code is exactly the same as it often will be, especially on x86. The issue is how we show in our source code that we have lock-free racy accesses to variables and that we are aware of this and have applied all the necessary atomic operations and/or orderAccess operations to ensure the races result in correct operation of the code. Unfortunately that is not currently feasible  due to the way this code has been written. I think fixing that aspect will need to be a future RFE.

I had already submitted tiers 1-3 with no problem, but have now submitted tiers 4-7.

I still need to re-examine the RCU code to understand the broader context here.

Thanks.

Reply:
[Not a review, just a drive-by comment.]
> We don't use class types in lock-free algorithms so don't have a means to
> express atomic updates to objects this way."

That isn't true. `oop` (when `CHECK_UNHANDLED_OOPS`) and `markWord` are the
first two that come to mind (there are some others). This is accomplished by
specializing PrimitiveConversions::Translate<> for these classes.